core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateCommandHandlerImpl.java

```diff
@@ -0,0 +1,271 @@
+import org.springframework.util.Assert;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Component to provide basic functionality around handling SolrUpdateCommands.
+ * @author Kelly Tisdell
+ *
+ */
+public abstract class AbstractSolrIndexUpdateCommandHandlerImpl implements SolrIndexUpdateCommandHandler {
+    
+    private static final Log LOG = LogFactory.getLog(AbstractSolrIndexUpdateCommandHandlerImpl.class);
+    
+    private final String commandGroup;
+    
+    public AbstractSolrIndexUpdateCommandHandlerImpl(String commandGroup) {
+        Assert.notNull(commandGroup, "Command group cannot be null.");
+        this.commandGroup = commandGroup.trim();
+        Assert.hasText(this.commandGroup, "Command group must not be empty and should not contain white spaces.");
+    }
+    
+    @Override
+    public String getCommandGroup() {
+        return commandGroup;
+    }
+
+    /**
+     * By default, this will update the foreground collection.  Deletes, if available, will be applied first.  Then, updates. This should be considered an autonomous method.
+     * Do not use this to make incremental updates within the scope of a larger update process because this will apply commits, by default.
+     * 
+     * @param command
+     * @throws ServiceException
+     */
+    protected void executeCommandInternal(IncrementalUpdateCommand command) throws ServiceException {
+        executeCommandInternal(command, getForegroundCollectionName());
+    }
+    
+    /**
+     * This will apply updates in the specified collection, and will commit, when finished, if no errors occur. 
+     * Deletes, if available, will be applied first.  Then, updates.  This should be considered an autonomous method.
+     * Do not use this to make incremental updates within the scope of a larger update process because this will apply commits, by default.
+     * 
+     * @param command
+     * @param collectionName
+     * @throws ServiceException
+     */
+    protected void executeCommandInternal(IncrementalUpdateCommand command, String collectionName) throws ServiceException {
+        Assert.notNull(command, "The command cannot be null.");
+        Assert.notNull(collectionName, "The collection name cannot be null.");
+        
+        boolean changeMade = false;
+        try {
+            
+            if (command.getDeleteQueries() != null && ! command.getDeleteQueries().isEmpty()) {
+                deleteByQueries(collectionName, command.getDeleteQueries());
+                changeMade = true;
+            }
+            
+            if (command.getSolrInputDocuments() != null && ! command.getSolrInputDocuments().isEmpty()) {
+                addDocuments(collectionName, command.getSolrInputDocuments());
+                changeMade = true;
+            }
+            
+            try {
+                if (changeMade) {
+                    commit(collectionName, true, true, false);
```

Comment:
While I agree that committing all deletes & additions at once is the right idea, there is potential for a huge commit if the command holds a large number of deletions and/or additions. 

Splitting into sets of smaller commits is likely the wrong idea since rollback scenarios get ugly quickly.

The best answer may be to enforce a limit on the number of deletions & additions per command. From there, it would be the responsibility of the caller to provide smaller commands.

Of course, this is assuming that there's a relatively low threshold for the max number of items that can be committed before seeing a significant performance degradation.